Encapsulate g_jscConfig and g_wtfConfig a tiny bit

geoffreygaren opened the pull request:
#### 99002f45c0c0b19808461a325ca0aacfdcdc9b16
<pre>
Encapsulate g_jscConfig and g_wtfConfig a tiny bit
<a href="https://bugs.webkit.org/show_bug.cgi?id=282673">https://bugs.webkit.org/show_bug.cgi?id=282673</a>
<a href="https://rdar.apple.com/139334844">rdar://139334844</a>

Reviewed by Chris Dumez.

These macros trigger bounds safety warnings in lots of places.

They are indeed written in an unsafe style, and we should fix that; but callers
don&apos;t pass indexing parameters into these macros, so warning about them at call
sites isn&apos;t useful.

I did one test removal of a warning suppression in this patch, to prove it works.

* Source/JavaScriptCore/runtime/JSCConfig.h:
(JSC::addressOfJSCConfig):
* Source/WTF/wtf/WTFConfig.h:
(WTF::addressOfWTFConfig):
* Source/WebKit/GPUProcess/EntryPoint/Cocoa/XPCService/GPUServiceEntryPoint.mm:
(GPU_SERVICE_INITIALIZER):

Canonical link: <a href="https://commits.webkit.org/286222@main">https://commits.webkit.org/286222@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0cdca0248369e468b4b470fb116e939a75ddbca

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75203 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79662 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26459 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63782 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2427 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59019 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17261 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39394 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46557 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22097 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24785 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68350 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81144 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74466 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2533 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67270 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2683 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64601 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66559 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10506 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8671 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96734 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11608 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2494 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5298 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21137 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2519 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3448 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2528 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/JavaScriptCore/runtime/JSCConfig.h:
```diff
@@ -123,7 +123,11 @@ constexpr size_t alignmentOfJSCConfig = std::alignment_of<JSC::Config>::value;
-#define g_jscConfig (*bitwise_cast<JSC::Config*>(&g_wtfConfig.spaceForExtensions))
+// Workaround to localize bounds safety warnings to this file.
+// FIXME: Use real types to make materializing JSC::Config* bounds-safe and type-safe.
+inline Config* addressOfJSCConfig() { return bitwise_cast<Config*>(&g_wtfConfig.spaceForExtensions); }
```
Can we remove some of the `WTF_ALLOW_UNSAFE_BUFFER_USAGE` statements in WebKit/ now? Several remain only to wrap `g_jscConfig` & `g_wtfConfig` iirc.

geoffreygaren replied:
Yes, I think we can. Why don't I add a test removal of one of them to this patch.

geoffreygaren replied:
OK, added.

geoffreygaren committed:
Encapsulate g_jscConfig and g_wtfConfig a tiny bit
https://bugs.webkit.org/show_bug.cgi?id=282673
rdar://139334844

Reviewed by Chris Dumez.

These macros trigger bounds safety warnings in lots of places.

They are indeed written in an unsafe style, and we should fix that; but callers
don't pass indexing parameters into these macros, so warning about them at call
sites isn't useful.

I did one test removal of a warning suppression in this patch, to prove it works.

* Source/JavaScriptCore/runtime/JSCConfig.h:
(JSC::addressOfJSCConfig):
* Source/WTF/wtf/WTFConfig.h:
(WTF::addressOfWTFConfig):
* Source/WebKit/GPUProcess/EntryPoint/Cocoa/XPCService/GPUServiceEntryPoint.mm:
(GPU_SERVICE_INITIALIZER):

Canonical link: https://commits.webkit.org/286222@main
